frontend/src/metabase/query_builder/components/view/QuestionDataSource.jsx

```diff
@@ -172,16 +173,24 @@ function getDataSourceParts({ question }) {
   }
 
   if (table) {
-    if (!isStructuredQuery) {
+    if (!isStructured) {
       return {
         name: table.displayName(),
       };
     }
 
     const allTables = [
       table,
-      ...query.joins().map(j => j.joinedTable()),
-    ].filter(Boolean);
+      ...Lib.joins(query, -1)
+        .map(join => Lib.pickerInfo(query, Lib.joinedThing(query, join)))
+        .map(pickerInfo => {
+          if (pickerInfo?.tableId != null) {
+            return metadata.table(pickerInfo.tableId);
+          } else if (pickerInfo?.cardId != null) {
+            return metadata.table(getQuestionVirtualTableId(pickerInfo.cardId));
+          }
```

Comment:
It's easier to read functions when:
- there's no `else` after `return`
- either all or none paths have `return` statements

```suggestion
          }
          
          if (pickerInfo?.cardId != null) {
            return metadata.table(getQuestionVirtualTableId(pickerInfo.cardId));
          }
          
          return undefined;
```

Reply:
04de16fbb278fa9810f30dcccea93472296ceaaa